File: solutions/2026-08-12/javascript/2026-08-12-length-of-longest-subarray-with-at-most-k-frequency.js
/**
 * @summary Finds the length of the longest subarray where each element's frequency is at most k.
 * @link https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency/
 * @approach This problem can be efficiently solved using the sliding window technique.
 * We maintain a window `[left, right]` and a frequency map (or hash table) to store the count of each element within the current window.
 * We expand the window by moving the `right` pointer. For each element `nums[right]`, we increment its count in the frequency map.
 * If the frequency of `nums[right]` exceeds `k`, it means the current window is no longer "good". To make it good again, we shrink the window from the left by moving the `left` pointer. As we shrink, we decrement the count of `nums[left]` in the frequency map. We continue shrinking until the frequency of `nums[right]` (or any element that caused the violation) becomes less than or equal to `k`.
 * At each step where the window is "good" (i.e., all element frequencies are <= k), we update the maximum length of the good subarray found so far, which is `right - left + 1`.
 *
 * Time Complexity: O(N), where N is the length of `nums`. Both `left` and `right` pointers traverse the array at most once. Hash map operations (insertion, deletion, lookup) take O(1) on average.
 * Space Complexity: O(M), where M is the number of distinct elements in `nums`. In the worst case, all elements are distinct, and the hash map stores all of them. However, since the problem statement implies that the number of distinct elements could be up to N, the space complexity can also be considered O(N) in the worst case if all elements are unique and `k` is large. Given the constraints, the maximum number of distinct elements is at most `nums.length`.
 */

/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
const longestSubarray = function(nums, k) {
    // Initialize a frequency map to store counts of elements in the current window.
    const freqMap = new Map();
    // Initialize the left pointer of the sliding window.
    let left = 0;
    // Initialize the maximum length of a good subarray found so far.
    let maxLength = 0;

    // Iterate through the array with the right pointer of the sliding window.
    for (let right = 0; right < nums.length; right++) {
        // Get the current element at the right pointer.
        const currentElement = nums[right];

        // Update the frequency of the current element in the map.
        // If the element is not in the map, initialize its count to 0 before incrementing.
        freqMap.set(currentElement, (freqMap.get(currentElement) || 0) + 1);

        // Check if the frequency of the current element exceeds k.
        // If it does, we need to shrink the window from the left to make it "good" again.
        while (freqMap.get(currentElement) > k) {
            // Get the element at the left pointer.
            const leftElement = nums[left];

            // Decrement the frequency of the element at the left pointer.
            freqMap.set(leftElement, freqMap.get(leftElement) - 1);

            // If the frequency of the element at the left pointer becomes 0, remove it from the map
            // to keep the map clean and efficient.
            if (freqMap.get(leftElement) === 0) {
                freqMap.delete(leftElement);
            }

            // Move the left pointer to the right, effectively shrinking the window.
            left++;
        }

        // At this point, the current window [left, right] is "good" because
        // the frequency of all elements within it is less than or equal to k.
        // Calculate the length of the current good subarray and update maxLength if it's larger.
        maxLength = Math.max(maxLength, right - left + 1);
    }

    // Return the maximum length of a good subarray found.
    return maxLength;
};
```

File: solutions/2026-08-12/java/2026-08-12-length-of-longest-subarray-with-at-most-k-frequency.java
// Problem: Length of Longest Subarray With at Most K Frequency
// Link: https://leetcode.com/problems/length-of-longest-subarray-with-at-most-k-frequency/
//
// Approach:
// This problem can be solved efficiently using the sliding window technique.
// We maintain a window defined by two pointers, `left` and `right`.
// We also use a hash map (`freqMap`) to store the frequency of each element within the current window.
// As we expand the window by moving the `right` pointer, we update the frequency of the current element.
// If, at any point, the frequency of an element exceeds `k`, it means the current window is no longer "good".
// To make the window "good" again, we shrink it from the left by moving the `left` pointer.
// While shrinking, we decrement the frequency of the element at the `left` pointer.
// We continue shrinking until the element's frequency becomes `k` or less, making the window "good" again.
// In each step where the window is "good" (i.e., all element frequencies are <= k), we update the `maxLength`
// by taking the maximum of its current value and the current window size (`right - left + 1`).
// This process ensures that we find the longest possible good subarray.
//
// Time Complexity: O(N), where N is the length of the `nums` array.
// Both `left` and `right` pointers traverse the array at most once.
// Hash map operations (get, put, remove) take O(1) on average.
//
// Space Complexity: O(M), where M is the number of distinct elements in `nums`.
// In the worst case, all elements are distinct, and the hash map will store all of them.
// M can be at most N.

import java.util.HashMap;
import java.util.Map;

class Solution {
    public int maxLength(int[] nums, int k) {
        // Initialize the frequency map to store counts of elements in the current window.
        Map<Integer, Integer> freqMap = new HashMap<>();
        // Initialize the left pointer of the sliding window.
        int left = 0;
        // Initialize the maximum length found so far.
        int maxLength = 0;

        // Iterate through the array with the right pointer to expand the window.
        for (int right = 0; right < nums.length; right++) {
            // Get the current element at the right pointer.
            int currentElement = nums[right];
            // Increment the frequency of the current element in the map.
            // If the element is not present, its frequency becomes 1.
            freqMap.put(currentElement, freqMap.getOrDefault(currentElement, 0) + 1);

            // Check if the frequency of the current element exceeds `k`.
            // If it does, we need to shrink the window from the left.
            while (freqMap.get(currentElement) > k) {
                // Get the element at the left pointer.
                int leftElement = nums[left];
                // Decrement the frequency of the element at the left pointer.
                freqMap.put(leftElement, freqMap.get(leftElement) - 1);

                // If the frequency of the left element becomes 0, remove it from the map.
                // This is an optimization, but not strictly necessary for correctness.
                if (freqMap.get(leftElement) == 0) {
                    freqMap.remove(leftElement);
                }

                // Move the left pointer to the right to shrink the window.
                left++;
            }

            // After ensuring the current window is "good" (all frequencies <= k),
            // update the maximum length found so far.
            // The current window length is `right - left + 1`.
            maxLength = Math.max(maxLength, right - left + 1);
        }

        // Return the maximum length of a good subarray.
        return maxLength;
    }
}
